test/test_torch.py

```diff
@@ -6987,80 +6988,74 @@ def test_logical_and(self, device):
+            mask = torch.randint(0, 2, X.shape, dtype=torch.bool, device=device)
+            X[mask] = nan
 
-        res1 = m1.clone()
-        res1.clamp_(min_val, max_val)
-        res2 = m1.clone()
-        for i in iter_indices(res2):
-            res2[i] = max(min_val, min(max_val, res2[i]))
-        self.assertEqual(res1, res2)
+        if isinstance(min_vals, torch.Tensor):
+            min_vals = min_vals.cpu().numpy()
 
-        out = m1.clone()
-        torch.clamp(m1, min=min_val, max=max_val, out=out)
-        self.assertEqual(out, res1)
+        if isinstance(max_vals, torch.Tensor):
+            max_vals = max_vals.cpu().numpy()
 
-        res1 = torch.clamp(m1, min=min_val)
-        res2 = m1.clone()
-        for i in iter_indices(res2):
-            res2[i] = max(min_val, res2[i])
-        self.assertEqual(res1, res2)
+        # Use NumPy implementation as reference
+        X_clamped = torch.tensor(np.clip(X.cpu().numpy(), a_min=min_vals, a_max=max_vals))
+        return X, X_clamped
 
-        torch.clamp(m1, min=min_val, out=out)
-        self.assertEqual(out, res1)
+    def test_clamp(self, device):
+        # Create min/max argument product
+        args = product((torch.linspace(-1, 1, 100, device=device), -1, None),
+                       (torch.linspace(-1, 1, 100, device=device), 1, None))
 
-        res1 = torch.clamp(m1, max=max_val)
-        res2 = m1.clone()
-        for i in iter_indices(res2):
-            res2[i] = min(max_val, res2[i])
-        self.assertEqual(res1, res2)
+        for min_val, max_val in args:
+            if min_val is None and max_val is None:
+                continue
 
-        torch.clamp(m1, max=max_val, out=out)
-        self.assertEqual(out, res1)
+            X, Y_expected = self.generate_clamp_baseline(device, min_val, max_val)
 
-        # if the tensor contains nan case
-        test_tens = torch.tensor([nan], device=device)
+            Y_actual = torch.clamp(X, min_val, max_val)
+            self.assertEqual(Y_expected, Y_actual)
 
-        res1 = test_tens.clone()
-        res1.clamp_(min_val, max_val)
-        res2 = test_tens.clone()
-        for i in iter_indices(res2):
-            res2[i] = max(min(res2[i], max_val), min_val)
-        self.assertEqual(torch.isnan(res1), torch.isnan(res2))
+            Y_out = torch.empty_like(X)
+            torch.clamp(X, min=min_val, max=max_val, out=Y_out)
+            self.assertEqual(Y_expected, Y_out)
 
-        out = test_tens.clone()
-        torch.clamp(test_tens, min=min_val, max=max_val, out=out)
-        self.assertEqual(torch.isnan(out), torch.isnan(res1))
+            X.clamp_(min_val, max_val)
+            self.assertEqual(Y_expected, X)
 
-        res1 = torch.clamp(test_tens, min=min_val)
-        res2 = test_tens.clone()
-        for i in iter_indices(res2):
-            res2[i] = max(res2[i], min_val)
-        self.assertEqual(torch.isnan(res1), torch.isnan(res2))
+    def test_clamp_raises_arg_errors(self, device):
```

Comment:
case for clamp with complex inputs failing, too?

Reply:
Yes, it fails, throwing `clamp does not support complex inputs`, but I didn't add a test for that as I believe it would be too verbose.